crypto/fipsmodule/curve25519/curve25519.c

```diff
@@ -135,16 +160,132 @@ void ED25519_keypair(uint8_t out_public_key[ED25519_PUBLIC_KEY_LEN],
+}
+
+int ED25519ctx_sign_no_self_test(
+    uint8_t out_sig[ED25519_SIGNATURE_LEN], const uint8_t *message,
+    size_t message_len, const uint8_t private_key[ED25519_PRIVATE_KEY_LEN],
+    const uint8_t *context, size_t context_len) {
+  return ed25519_sign_internal(ED25519CTX_ALG, out_sig, message, message_len,
+                               private_key, context, context_len);
+}
+
+int ED25519ph_sign(uint8_t out_sig[ED25519_SIGNATURE_LEN],
+                   const uint8_t *message, size_t message_len,
+                   const uint8_t private_key[ED25519_PRIVATE_KEY_LEN],
+                   const uint8_t *context, size_t context_len) {
+  boringssl_ensure_eddsa_self_test();
+  return ED25519ph_sign_no_self_test(out_sig, message, message_len, private_key,
+                                     context, context_len);
+}
+
+int ED25519ph_sign_no_self_test(
+    uint8_t out_sig[ED25519_SIGNATURE_LEN], const uint8_t *message,
+    size_t message_len, const uint8_t private_key[ED25519_PRIVATE_KEY_LEN],
+    const uint8_t *context, size_t context_len) {
+  uint8_t digest[SHA512_DIGEST_LENGTH] = {0};
+  FIPS_service_indicator_lock_state();
+  SHA512_CTX ctx;
+  SHA512_Init(&ctx);
+  SHA512_Update(&ctx, message, message_len);
+  SHA512_Final(&digest[0], &ctx);
+  FIPS_service_indicator_unlock_state();
+  return ED25519ph_sign_digest_no_self_test(out_sig, digest, private_key,
+                                            context, context_len);
+}
+
+int ED25519ph_sign_digest(uint8_t out_sig[ED25519_SIGNATURE_LEN],
+                          const uint8_t digest[SHA512_DIGEST_LENGTH],
+                          const uint8_t private_key[ED25519_PRIVATE_KEY_LEN],
+                          const uint8_t *context, size_t context_len) {
+  boringssl_ensure_eddsa_self_test();
+  return ED25519ph_sign_digest_no_self_test(out_sig, digest, private_key,
+                                            context, context_len);
+}
+
+int ED25519ph_sign_digest_no_self_test(
+    uint8_t out_sig[ED25519_SIGNATURE_LEN],
+    const uint8_t digest[SHA512_DIGEST_LENGTH],
+    const uint8_t private_key[ED25519_PRIVATE_KEY_LEN],
+    const uint8_t *context, size_t context_len) {
+  return ed25519_sign_internal(ED25519PH_ALG, out_sig, &digest[0],
+                               SHA512_DIGEST_LENGTH, private_key, context,
+                               context_len);
+}
+
+static int dom2(ed25519_algorithm_t alg, uint8_t buffer[MAX_DOM2_SIZE],
+                size_t *buffer_len, const uint8_t *context,
+                size_t context_len) {
+  GUARD_PTR(buffer_len);
+  *buffer_len = 0;
+
+  uint8_t phflag = 0;
+
+  switch (alg) {
+    case ED25519_ALG:
+      // Per rfc8032:
+      // For Ed25519, dom2(f,c) is the empty string.  The phflag value is
+      // irrelevant.  The context (if present at all) MUST be empty.
+      return context_len == 0;
+    case ED25519CTX_ALG:
+      // Per rfc8032:
+      // For Ed25519ctx, phflag=0.  The context input SHOULD NOT be empty.
```

Comment:
Isn't SHOULD optional? https://datatracker.ietf.org/doc/html/rfc2119#section-3 or are you saying for AWS-LC's implementation it MUST NOT be empty?

Reply:
It seems the SHOULD is treated as a MUST NOT, as the OpenSSL APIs for Ed25519ctx also require the context to be non-empty. It was also called out by Nevine as a must not.